subsys/net/lib/dns/dns_cache.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2024 Endress+Hauser AG
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/net/dns_resolve.h>
+#include "dns_cache.h"
+
+LOG_MODULE_REGISTER(net_dns_cache, 4);
+
+static void dns_cache_clean(struct dns_cache const *cache);
+
+int dns_cache_flush(struct dns_cache *cache)
+{
+	k_mutex_lock(cache->mut, K_FOREVER);
+	for (size_t i = 0; i < cache->size; i++) {
+		cache->entries[i].in_use = false;
+	}
+	k_mutex_unlock(cache->mut);
+
+	return 0;
+}
+
+int dns_cache_add(struct dns_cache *cache, char const *query, struct dns_addrinfo const *addrinfo,
+		  uint32_t ttl)
+{
+	int64_t oldest_entry_uptime = INT64_MAX;
+	size_t index_to_replace = 0;
+	bool found_empty = false;
+
+	if (cache == NULL || query == NULL || addrinfo == NULL || ttl == 0) {
+		return -EINVAL;
+	}
+
+	if (strlen(query) >= DNS_MAX_NAME_LEN) {
+		LOG_WRN("Query string to big to be processed %u >= DNS_MAX_NAME_LEN",
+			strlen(query));
+		return -EINVAL;
+	}
+
+	k_mutex_lock(cache->mut, K_FOREVER);
+
+	LOG_DBG("ADD \"%s\" with TTL %" PRIu32, query, ttl);
+
+	dns_cache_clean(cache);
```

Comment:
True but then we would need a loop more as soon as the cache is full every time we add something (without calling find). Do you think this would still be more efficient? Or is there something I am missing here?
```c
        // Additional loop to check if space is available
	for (size_t i = 0; i < cache->size; i++) {
		if (!cache->entries[i].in_use) {
			index_to_replace = i;
			found_empty = true;
		}
	}

	if(!found_empty) {
		dns_cache_clean(cache);
		for (size_t i = 0; i < cache->size; i++) {
			if (!cache->entries[i].in_use) {
				index_to_replace = i;
				found_empty = true;
				break;
			} else if (cache->entries[i].creation_uptime < oldest_entry_uptime) {
				index_to_replace = i;
				oldest_entry_uptime = cache->entries[i].creation_uptime;
			}
		}
	}
```